File: Spark/Pyspark/orderRevenueAndCountPerDay.py
from pyspark import SparkContext, SparkConf

def parseOrders(rec):
    parts = rec.split(",")
    return (int(parts[0]), parts[1], int(parts[2]), parts[3]  )


def parseOrderItems(rec):
    parts = rec.split(",")
    return (int(parts[0]), int(parts[1]), int(parts[2]), int(parts[3]), float(parts[4]), float(parts[5])  )

conf = SparkConf()
conf.setMaster("local")
conf.setAppName("OrderRevenueCountPerDay")

sc = SparkContext(conf=conf)

path = "/user/hive/warehouse/retail_db.db"

orders = sc.textFile(path + "/orders").map(lambda x: parseOrders(x))
orderItems = sc.textFile(path + "/order_items").map(lambda x: parseOrderItems(x))

# We cannot join the tuples together even if we wanted to join on first column because 
# spark will only take first column as key and the 2nd column from each dataset as value 
# and ignore the remaining fields from both datasets

ordersKV = orders.map(lambda x: (x[0], (x[1], x[2], x[3])))
orderItemsKV = orderItems.map(lambda x: (x[1], (x[0], x[2], x[3], x[4], x[5])))

# Join Demo.
ordersJoin = ordersKV.join(orderItemsKV)

# Generate a flat structure with order_date and orderId  as key and order_subtotal as value.
ordersByOrderIdDate = ordersJoin.map(lambda x : ( (x[0],x[1][0][0]), x[1][1][3]))

# Calculated totals of orders
orderTotalsPerOrder = ordersByOrderIdDate.reduceByKey( lambda acc,val: acc +val)

# Generate a flat structure with Date as key and subtotal as value
ordersByDate = orderTotalsPerOrder.map(lambda x: (x[0][1], x[1]))

# now again reduce by key (orderdate) to find total revenure and ordercount per day
# However since there is change in the input type (float) and output type (float, int) we need to use aggregate by key
#revenueAndOrderCountPerDay = ordersByDate.aggregateByKey( (0.0, 0)  ,  lambda acc, val : (acc[0] + val, acc[1] + 1), lambda acc,val: (acc[0] + val[0], acc[1] + val[1]) )

# OR we can use combineByKey
# The first parameter is a function which takes in a record (the first record from each partition is passed to this function.
# the Second parameter is the combiner function which takes in the tuple returned by first parameter and the 2nd Value, 3rd value and so on...
revenueAndOrderCountPerDay = ordersByDate.combineByKey( lambda t: (t, 1), lambda acc, val: (acc[0] + val, acc[1] + 1), lambda acc, val: (acc[0] + val[0], acc[1] + val[1]) )

temp = revenueAndOrderCountPerDay.take(100)
for x in temp: print(x)

File: Spark/Scala/orderRevenueAndCountPerDay.scala
import org.apache.spark._;
import org.apache.spark.rdd._;

val conf = new SparkConf().setMaster("local").setAppName("OrderRevenueAndCountPerDay Scala");

val sc = new SparkContext(conf)


val path = "/user/hive/warehouse/retail_db.db";


case class Order(order_id:Int, order_date:String, order_cust_id: Int, order_status:String)
def parseOrder(rec:String): Order ={
	
	val parts = rec.split(",");
	return Order(parts(0).toInt, parts(1), parts(2).toInt, parts(3))

}

case class OrderItem(order_item_id:Int, order_item_order_id: Int, order_item_prod_id: Int, order_item_quantity:Int, order_item_subtotal:Float, order_item_prod_price:Float);
def parseOrderItem(rec:String) : OrderItem = {

	val parts = rec.split(",");
	return OrderItem(parts(0).toInt, parts(1).toInt, parts(2).toInt, parts(3).toInt, parts(4).toFloat, parts(5).toFloat);
}

// Parse the records into corresponding objects
val orders = sc.textFile(path + "/orders").map(x => parseOrder(x));
val orderItems = sc.textFile(path + "/order_items").map(x => parseOrderItem(x));

// Generate KV pairs where K- order id, V = object
val ordersKV = orders.map(x => (x.order_id, x));
val orderItemsKV = orderItems.map(x => (x.order_item_order_id, x));

// Join on key
val ordersJoin = ordersKV.join(orderItemsKV);

// Generate a flat structure which contains K = order_id, order_date and V= subtotal
val ordersByOrderIdDate = ordersJoin.map(x => ((x._2._1.order_id,x._2._1.order_date), x._2._2.order_item_subtotal));

// Reduce by key to calculate orders total
val ordersTotal = ordersByOrderIdDate.reduceByKey( (acc,value) => acc + value);

// Generate the flat structure where K= order_date, V = totals
val ordersByDate = ordersTotal.map(x => (x._1._2, x._2))

// Again reduce by key to find total revenue and count per day. Since the input is just float (total) and output is revenue and count tuple(float, int) 
// we need to use aggregateByKey
val revenueAndCountPerDay = ordersByDate.aggregateByKey((0.0, 0) )( (acc, value) => (acc._1 + value, acc._2 + 1) , (acc, value) => (acc._1 + value._1, acc._2 + value._2));

revenueAndCountPerDay.take(10).foreach(println)


 
